pkg/middleware/auth.go

```diff
@@ -0,0 +1,175 @@
+/****************************************************************************
+ * Copyright 2020, Optimizely, Inc. and contributors                   *
+ *                                                                          *
+ * Licensed under the Apache License, Version 2.0 (the "License");          *
+ * you may not use this file except in compliance with the License.         *
+ * You may obtain a copy of the License at                                  *
+ *                                                                          *
+ *    http://www.apache.org/licenses/LICENSE-2.0                            *
+ *                                                                          *
+ * Unless required by applicable law or agreed to in writing, software      *
+ * distributed under the License is distributed on an "AS IS" BASIS,        *
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. *
+ * See the License for the specific language governing permissions and      *
+ * limitations under the License.                                           *
+ ***************************************************************************/
+
+// Package middleware //
+package middleware
+
+import (
+	"errors"
+	"fmt"
+	"net/http"
+	"strings"
+	"time"
+
+	"github.com/dgrijalva/jwt-go"
+)
+
+func getNumberFromJSON(val interface{}) int64 {
+	switch v := val.(type) {
+	case int64:
+		return v
+	case float64:
+		return int64(v)
+	}
+	return 0
+}
+
+// NoAuth is NoOp for auth
+type NoAuth struct{}
+
+// CheckToken returns no token and no error
+func (NoAuth) CheckToken(string) (*jwt.Token, error) {
+	return nil, nil
+}
+
+// Auth is the middleware for all REST API's
+type Auth struct {
+	Verifier
+
+	checkClaims map[string]struct{}
+}
+
+// Verifier checks token
+type Verifier interface {
+	CheckToken(string) (*jwt.Token, error)
+}
+
+// JWTVerifier checks token with JWT, implements Verifier
+type JWTVerifier struct {
+	secretKey string
+}
+
+// NewJWTVerifier creates JWTVerifier with secret key
+func NewJWTVerifier(secretKey string) JWTVerifier {
+	return JWTVerifier{secretKey: secretKey}
+}
+
+// CheckToken checks the token and returns it if it's valid
+func (c JWTVerifier) CheckToken(token string) (*jwt.Token, error) {
+	if token == "" {
+		return nil, errors.New("invalid token")
+	}
+
+	tk, err := jwt.Parse(token, func(token *jwt.Token) (interface{}, error) {
+		if _, ok := token.Method.(*jwt.SigningMethodHMAC); !ok {
+			return nil, errors.New("unexpected signing method")
+		}
+		return []byte(c.secretKey), nil
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	if !tk.Valid {
+		return nil, errors.New("invalid token")
+	}
+
+	return tk, nil
+}
+
+func (a Auth) verify(r *http.Request) (*jwt.Token, error) {
+
+	var token string
+
+	if values, ok := r.Header["Auth"]; ok && len(values) > 0 {
+		token = values[0]
+	}
+
+	if values, ok := r.Header["Jwt"]; ok && len(values) > 0 {
+		token = values[0]
+	}
```

Comment:
Are these standard header keys? Can we move to configuration as opposed to sniffing these out?

Reply:
I put here most popular and common prefixes for token checks